src/index.ts

```diff
@@ -13,80 +13,76 @@
-import {perftools} from './profile';
+const common: Common = require('@google-cloud/common');
 
-type HrTimeTuple = [number, number];
+// initConfig sets unset values in the configuration to the value retrieved from
+// environment variables, metadata, or the default value specified in
+// defaultConfig.
+// Returns rejected promise if value that must be set cannot be initialized.
+async function initConfig(config: Config): Promise<ProfilerConfig> {
+  config = common.util.normalizeArguments(null, config);
 
-const profiler = require('bindings')('cpu_profiler');
+  const envConfig = {
+    logLevel: process.env.GCLOUD_PROFILER_LOGLEVEL,
+    projectId: process.env.GCLOUD_PROJECT,
+    serviceContext: {
+      service: process.env.GAE_SERVICE,
+      version: process.env.GAE_VERSION,
+    }
+  };
 
-function timeToNanos(tuple: HrTimeTuple) {
-  return tuple[0] * 1e9 + tuple[1];
-}
-
-const durationMillis = 10 * 1000;
-const intervalMillis = 60 * 1000;
-let isActive = false;
+  let envSetConfig: Config = {};
+  if (process.env.hasOwnProperty('GCLOUD_PROFILER_CONFIG')) {
+    envSetConfig =
+        require(path.resolve(process.env.GCLOUD_PROFILER_CONFIG)) as Config;
+  }
 
-function profileInterval() {
-  assert(durationMillis <= intervalMillis);
-  const startDelay = (intervalMillis - durationMillis) * Math.random();
-  setTimeout(function() {
-    const startTime = Date.now();
-    const runName = 'cloud-profile-' + startTime;
-    profiler.startProfiling(runName, true);
-    isActive = true;
-    setTimeout(function() {
-      isActive = false;
-      const result = profiler.stopProfiling(runName);
-      const serialized = serialize(result, startTime * 1e6);
-      const writer = perftools.profiles.Profile.encode(serialized);
-      const buffer = writer.finish();
+  let normalizedConfig = extend(true, {}, defaultConfig, envConfig, config);
 
-      const outp = fs.createWriteStream(runName + '.pb.gz');
-      const inp = new stream.PassThrough();
-      inp.end(buffer);
-      inp.pipe(zlib.createGzip()).pipe(outp).on('close', function() {
-        setTimeout(
-            profileInterval, intervalMillis - startDelay - durationMillis)
-            .unref();
-      });
-    }, durationMillis).unref();
-  }, startDelay).unref();
-}
+  if (normalizedConfig.serviceContext.service === undefined) {
+    throw new Error('service name must be specified in the configuration');
+  }
 
-process.on('exit', function() {
-  if (isActive) {
-    profiler.stopProfiling();
+  // TODO: fetch instance and zone from metadata. This will require function to
+  // be asynchrous.
+  if (normalizedConfig.instance === undefined) {
+    normalizedConfig.instance = '';
   }
-});
-process.on('uncaughtException', function() {
-  if (isActive) {
-    profiler.stopProfiling();
+  if (normalizedConfig.zone === undefined) {
+    normalizedConfig.zone = '';
   }
-});
 
-profileInterval();
+  return normalizedConfig;
+}
+
+let profiler: Profiler|undefined = undefined;
```

Comment:
nit: Would `Profiler?` work?

Reply:
Doesn't seem to work.

Reply:
Specifically `let profiler: Profiler? = undefined;`  gave me this error:
64 let profiler: Profiler? = undefined;
       ~~~~~~~~

src/index.ts(64,5): error TS2322: Type 'undefined' is not assignable to type 'Profiler | null'.


64 let profiler: Profiler? = undefined;
                 ~~~~~~~~~

src/index.ts(64,15): error TS8020: JSDoc types can only be used inside documentation comments.


And  `let profiler?: Profiler = undefined;` gave me this error:

64 let profiler?: Profiler = undefined;
               ~

src/index.ts(64,13): error TS1005: ',' expected.


64 let profiler?: Profiler = undefined;
                ~

src/index.ts(64,14): error TS1134: Variable declaration expected.